### PR TITLE
[Region] Bufferize with one-shot-bufferize

### PR DESCRIPTION
**Pull request by tkarna**
- Remove region-bufferize pass and Region dialect conversion patterns
- Add BufferizableOpInterface extensions to Region dialect to enable bufferization via one-shot-bufferize
- Update region bufferize test that uses one-shot-bufferize

---

Please review these guidelines to help with the review process:
- [x] Have you provided a meaningful PR description?
- [x] Have you added a test, a reproducer, or a reference to an issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [x] If this PR is a work in progress, are you filing the PR as a draft?
- [x] Have you organized your commits logically and ensured each can be built by itself?

### CODE REVIEW

**Review by silee2 — APPROVED** [review]

File: lib/Dialect/Region/Transforms/BufferizableOpInterfaceImpl.cpp [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+  }
+
+  ::mlir::LogicalResult
+  bufferize(::mlir::Operation *op, ::mlir::RewriterBase &rewriter,
+            const ::mlir::bufferization::BufferizationOptions &options) const {
+    auto envOp = ::mlir::cast<region::EnvironmentRegionOp>(op);
+    if (noTensorsIn(envOp.getArgs()) && noTensorsIn(envOp.getResults())) {
+      // Nothing to do.
+      return ::mlir::success();
+    }
```
Comment (fschlimb) [inline lib/Dialect/Region/Transforms/BufferizableOpInterfaceImpl.cpp:87]: Is this needed? the docs say
```
         This method will never be called on ops that do not have at least one
         tensor operand/result.
```

File: lib/Dialect/Region/Transforms/BufferizableOpInterfaceImpl.cpp [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    return ::llvm::isa<::mlir::TensorType>(v.getType());
+  };
+  if (::llvm::any_of(values, isTensor))
+    return false;
+  return true;
+}
+
+/// Convert values to buffers. If a value is a tensor, get a buffer for it.
+::mlir::LogicalResult
+convertToBuffers(::mlir::ValueRange values,
```
Comment (fschlimb) [inline lib/Dialect/Region/Transforms/BufferizableOpInterfaceImpl.cpp:30]: Why do you need this? Shouldn't `getBuffer` be sufficient?

File: lib/Dialect/Region/Transforms/BufferizableOpInterfaceImpl.cpp [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+/// Bufferization of region.env_region_yield. Replaced with a new yield that
+/// operates on memrefs.
+struct EnvironmentRegionYieldOpInterface
+    : public ::mlir::bufferization::BufferizableOpInterface::ExternalModel<
+          EnvironmentRegionYieldOpInterface, region::EnvironmentRegionYieldOp> {
+  bool bufferizesToMemoryRead(
+      ::mlir::Operation *op, ::mlir::OpOperand &opOperand,
+      const ::mlir::bufferization::AnalysisState &state) const {
+    return true;
```
Comment (fschlimb) [inline lib/Dialect/Region/Transforms/BufferizableOpInterfaceImpl.cpp:126]: what is it reading?

Comment (tkarna) [reply lib/Dialect/Region/Transforms/BufferizableOpInterfaceImpl.cpp:126]: This implementation follows ` scf.yield` op bufferization: read is true, write is false etc.

File: lib/Dialect/Region/Transforms/BufferizableOpInterfaceImpl.cpp (hunk 1 above)
Comment (tkarna) [reply lib/Dialect/Region/Transforms/BufferizableOpInterfaceImpl.cpp:87]: Agreed, this is not needed.

**Review by fschlimb — APPROVED** [review]

**Pull request merged**